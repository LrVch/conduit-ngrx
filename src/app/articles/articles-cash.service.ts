import { Injectable } from '@angular/core';
import * as schema from './schema';
import { normalize } from 'normalizr';
import { Article } from '@app/core';

export interface ResponseArticles {
  articles: Article[];
  articlesCount: number;
}

export interface NormalizedArticlesResponse {
  entities: {
    articles: {
      [key: string]: Article;
    };
  };
  result: {
    articles: string[];
    articlesCount: number;
  };
}

@Injectable({
  providedIn: 'root'
})
export class ArticlesCashService {
  public static normalizeArticlesResponce(
    articles: ResponseArticles
  ): NormalizedArticlesResponse {
    return normalize(articles, schema.responceArticlesSchema);
  }

  constructor() {}
}
