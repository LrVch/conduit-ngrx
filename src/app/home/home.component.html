<section>
  <div class="banner text-center" *appShowAuthed="false">
    <div class="container">
      <h1>conduit</h1>
      <p>A place to share your Angular 2 knowledge.</p>
    </div>
  </div>

  <div class="content">

    <div class="container">
      <div class="row">
        <div class="col-md-9 col-sm-12">
          <app-content-tabs [type]="type$ | async"
                            [tabs]="tabs"
                            (selectedType)="onSelectedType($event)">
            <mat-chip *ngIf="(currentTag$ | async) as currentTag"
                      selected
                      color="primary">#{{ currentTag }}</mat-chip>
          </app-content-tabs>
        </div>
      </div>
    </div>

    <div class="container">
      <div class="row flex-column-reverse flex-lg-row">
        <div class="col-lg-9 col-sm-12">
          <div class="articles">
            <app-articles-list [isLoading]="loadingArticles$ | async"
                               [isErrorLoading]="errorLoadingArticles$ | async"
                               (favorited)="onFavorited($event)"
                               [articlesList]="atricles$ | async"></app-articles-list>
          </div>
        </div>

        <div class="col-lg-3 col-sm-12">
          <div class="tags-wtapper">
            <app-tags [tags]="tags$ | async"
                      [isLoading]="loadingTags$ | async"
                      [isErrorLoading]="errorLoadingTags$ | async"
                      [currentTag]="currentTag$ | async"
                      (selecteTag)="onSelectedTag($event)"></app-tags>
          </div>
        </div>
      </div>
    </div>

    <div class="container">
      <div class="row">
        <div class="col-md-9 col-sm-12">
          <mat-paginator [hidden]="(articlesCount$ | async) < 1 || (loadingArticles$ | async)"
                         [length]="articlesCount$ | async"
                         (page)="onPageChange($event)"
                         [pageSize]="limit$ | async"
                         [showFirstLastButtons]="true"
                         [pageSizeOptions]="pageSizeOptions">
          </mat-paginator>
        </div>
      </div>
      <div class="pagination">
      </div>
    </div>
  </div>
</section>