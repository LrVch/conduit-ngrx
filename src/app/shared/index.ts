export * from './article/article.component';
export * from './articles-list/articles-list.component';
export * from './content-tabs/content-tabs.component';
export * from './tags/tags.component';
export * from './favorite/favorite.component';
export * from './layout';
export * from './shared.module';
export * from './utils';
export * from './list-error/list-error.component';
export * from './show-authed.directive';
export * from './user-info/user-info.component';
export * from './follow-button/follow-button.component';
export * from './tag-list/tag-list.component';
export * from './comments/comments.component';
export * from './base-from/base-from.component';
