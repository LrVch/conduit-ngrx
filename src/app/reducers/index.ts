import {
  ActionReducerMap,
  MetaReducer
} from '@ngrx/store';
import { environment } from '@env/environment';
import { AuthState, authReducer } from '@app/auth/auth.reducer';
import { ArticlesState, articlesReducer } from '@app/articles/articles.reducer';
import { storeFreeze } from 'ngrx-store-freeze';
import { routerReducer, RouterReducerState } from '@ngrx/router-store';
import { ArticlesConfigState, articlesConfigReducer } from '@app/articles/articlesConfig.reducer';
import { ProfileState, profileReducer } from '@app/profile/profile.reducer';
import { layoutReducer, LayoutState } from '@app/layout/layout.reducer';
import { EditorState, editorReducer } from '@app/editor/editor.reducer';
import { ArticleState, articleReducer } from '@app/article/article.reducer';
import { AppSettingsState } from '@app/core/models/app-settings.model';
import { appSettingsReducer } from '@app/appSettings/app-settings.reducer';
import { debug } from './meta-reducers/debug.reducer';

import { initStateFromLocalStorage } from './meta-reducers/init-state-from-local-storage.reducer';


export interface AppState {
  auth: AuthState;
  articles: ArticlesState;
  router: RouterReducerState;
  articlesConfig: ArticlesConfigState;
  profile: ProfileState;
  layout: LayoutState;
  editor: EditorState;
  article: ArticleState;
  appSettings: AppSettingsState;
}

export const reducers: ActionReducerMap<AppState> = {
  auth: authReducer,
  articles: articlesReducer,
  router: routerReducer,
  articlesConfig: articlesConfigReducer,
  profile: profileReducer,
  layout: layoutReducer,
  editor: editorReducer,
  article: articleReducer,
  appSettings: appSettingsReducer
};


export const metaReducers: MetaReducer<AppState>[] = [initStateFromLocalStorage];

if (!environment.production) {
  metaReducers.unshift(storeFreeze);
}
