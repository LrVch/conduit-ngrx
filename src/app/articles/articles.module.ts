import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { StoreModule } from '@ngrx/store';
import * as fromArticles from './articles.reducer';
import * as fromArticlesConfig from './articlesConfig.reducer';
import { EffectsModule } from '@ngrx/effects';
import { ArticlesEffects } from './articles.effects';
import { SharedModule } from '@app/shared';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment';

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forFeature('articles', fromArticles.articlesReducer),
    StoreModule.forFeature('articlesConfig', fromArticlesConfig.articlesConfigReducer),
    EffectsModule.forFeature([ArticlesEffects]),
    SharedModule,
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
    }),
  ],
  declarations: []
})
export class ArticlesModule { }

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.i18nPrefix}/assets/i18n/`,
    '.json'
  );
}
