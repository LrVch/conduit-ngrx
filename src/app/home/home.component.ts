import { Component, OnInit, AfterViewChecked } from '@angular/core';
import { Observable, combineLatest } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import {
  LoadTagsRequest,
  LoadArticlesRequest,
  SetTypeOfFeed,
  ResetConfig,
  SetTag,
  SetLimit,
  SetOffset,
  ToggleArticleFavoriteRequest,
  SetPageIndex,
  ClearReturnArticlesConfig
} from '../articles/articles.actions';
import {
  selectArticlesTags,
  selectArticlesTagsLoading,
  selectArticlesTagsFailLoading,
  getArticlesLimit,
  selectArticlesItems,
  getArticlesCurrentTag,
  getArticlesFiltersType,
  selectArticlesCount,
  selectArticlesLoading,
  getArticlesPageIndex,
  selectArticlesReturnConfig
} from '../articles/articles.selectors';
import { Article } from '../core';
import { selectAuthLoggedIn } from '../auth/auth.selectors';
import { first, map, mergeAll, tap, startWith } from 'rxjs/operators';
import { LogoutAction } from '../auth/auth.actions';
import { ArticlesConfigState } from '../articles/articlesConfig.reducer';
import { TranslateService } from '@ngx-translate/core';

@Component({
  templateUrl: './home.component.html'
})
export class HomeComponent implements OnInit {
  constructor(
    private store: Store<AppState>,
    private translateService: TranslateService
  ) { }

  isAuthenticated$: Observable<boolean>;
  loadingTags$: Observable<boolean>;
  errorLoadingTags$: Observable<boolean>;
  tags$: Observable<Array<string>>;
  limit$: Observable<number>;
  pageIndex$: Observable<number>;
  atricles$: Observable<Article[]>;
  currentTag$: Observable<string>;
  type$: Observable<string>;
  tabs$: Observable<{ title: string, value: string }[]>;
  pageSizeOptions = [5, 10, 25, 100];
  isLoggegId: boolean;
  articlesCount$: Observable<number>;
  loadingArticles$: Observable<boolean>;
  errorLoadingArticles$: Observable<boolean>;

  ngOnInit() {
    combineLatest(
      this.store.pipe(select(selectArticlesReturnConfig)),
      this.store.pipe(select(selectAuthLoggedIn)),
    ).pipe(first()).subscribe(([config, isLoggegIn]: [ArticlesConfigState, boolean]) => {
      this.store.dispatch(new ClearReturnArticlesConfig());
      this.store.dispatch(new ResetConfig());

      if (config && isLoggegIn) {
        const { tag, offset, limit } = config.filters;
        this.store.dispatch(new SetTypeOfFeed({ type: config.type }));
        this.store.dispatch(new SetLimit({ limit }));
        this.store.dispatch(new SetOffset({ offset }));
        this.store.dispatch(new SetPageIndex({ pageIndex: offset / limit }));
        this.store.dispatch(new SetTag({ tag }));
      } else {
        if (isLoggegIn) {
          this.store.dispatch(new SetTypeOfFeed({ type: 'feed' }));
        }
      }
    });

    this.tabs$ =
      this.translateService.onLangChange.pipe(
        startWith(this.translateService.get(['conduit.home.tabGlobal', 'conduit.home.tabFeed'])),
        map(() => this.translateService.get(['conduit.home.tabGlobal', 'conduit.home.tabFeed'])),
        mergeAll(),
        map(result => {
          return [{ title: result['conduit.home.tabGlobal'], value: 'all' }, { title: result['conduit.home.tabFeed'], value: 'feed' }];
        })
      );

    this.store.dispatch(new LoadTagsRequest);
    this.store.dispatch(new LoadArticlesRequest());

    this.tags$ = this.store.pipe(select(selectArticlesTags));
    this.currentTag$ = this.store.pipe(select(getArticlesCurrentTag));
    this.loadingTags$ = this.store.pipe(select(selectArticlesTagsLoading));
    this.errorLoadingTags$ = this.store.pipe(select(selectArticlesTagsFailLoading));

    this.type$ = combineLatest(this.store.pipe(select(getArticlesFiltersType)), this.currentTag$).pipe(
      map(([type, tag]) => {
        return tag ? '' : type;
      })
    );

    this.atricles$ = this.store.pipe(select(selectArticlesItems));
    this.limit$ = this.store.pipe(select(getArticlesLimit));
    this.pageIndex$ = this.store.pipe(select(getArticlesPageIndex));
    this.articlesCount$ = this.store.pipe(select(selectArticlesCount));
    this.loadingArticles$ = this.store.pipe(select(selectArticlesLoading));
  }

  onSelectedTag(tag: string): void {
    this.store.dispatch(new ResetConfig());
    this.store.dispatch(new SetTag({ tag }));
    this.store.dispatch(new LoadArticlesRequest());
  }

  onSelectedType(type: string): void {
    this.store.pipe(select(selectAuthLoggedIn)).pipe(first()).subscribe(isLoggegId => {
      this.store.dispatch(new ResetConfig());
      this.store.dispatch(new SetTypeOfFeed({ type }));
      if (type === 'feed' && !isLoggegId) {
        this.store.dispatch(new LogoutAction());
        return;
      }
      this.store.dispatch(new LoadArticlesRequest());
    });
  }

  onPageChange(page): void {
    const { pageIndex, pageSize } = page;
    const limit = pageSize;
    const offset = limit * pageIndex;
    this.store.dispatch(new SetLimit({ limit }));
    this.store.dispatch(new SetPageIndex({ pageIndex }));
    this.store.dispatch(new SetOffset({ offset }));
    this.store.dispatch(new LoadArticlesRequest());
  }

  onFavorited(article: Article): void {
    this.store.dispatch(new ToggleArticleFavoriteRequest({ article }));
  }
}
