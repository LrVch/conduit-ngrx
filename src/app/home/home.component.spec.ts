import { async, ComponentFixture } from '@angular/core/testing';
import { ConfigureFn, configureTests, getUser, getProfile, getArticle } from '@app/lib/testing';
import { HomeComponent } from './home.component';
import { DebugElement, Component, Output, EventEmitter, Input } from '@angular/core';
import { Errors, User, Profile, Article } from '@app/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { Store, StoreModule, combineReducers } from '@ngrx/store';
import * as ArticlesActions from '@app/articles/articles.actions';
import * as AuthActions from '@app/auth/auth.actions';
import * as fromRoot from '@app/reducers';
import * as fromArticles from '@app/articles/articles.reducer';
import * as fromAuth from '@app/auth/auth.reducer';
import { hot, cold } from 'jasmine-marbles';
import { ArticlesConfigState } from '@app/articles/articlesConfig.reducer';
import { of } from 'rxjs';
import { ShowAuthedDirective } from '@app/shared';


@Component({
  selector: 'app-tags',
  template: `<div id="selectTag" (click)="onSelectedTag()"></div>`
})
class TestHostTagsComponent {
  @Input('tags') tags: any;
  @Input('isErrorLoading') isErrorLoading: any;
  @Input('currentTag') currentTag: any;
  @Input('isLoading') isLoading: any;
  @Output() selecteTag = new EventEmitter<any>();

  onSelectedTag() {
    this.selecteTag.emit('tag');
  }
}

@Component({
  selector: 'app-content-tabs',
  template: `<div id="tabs" (click)="onSelectedType($event)"><ng-content></ng-content></div>`
})
class TestHostContentTabsComponent {
  @Input('type') type: any;
  @Input('tabs') tabs: any;
  @Output() selectedType = new EventEmitter();

  onSelectedType(event) {
    this.selectedType.emit(event.type);
  }
}

@Component({
  selector: 'app-articles-list',
  template: `<div id="favorite" (click)="onFavorited()"></div>`
})
class TestHostArticleListComponent {
  @Input('isLoading') isLoading: boolean;
  @Input('isErrorLoading') isErrorLoading: boolean;
  @Input('articlesList') articlesList: any;
  @Output() favorited = new EventEmitter();

  onFavorited() {
    this.favorited.emit(getArticle());
  }
}

@Component({
  selector: 'mat-chip',
  template: `<ng-content></ng-content>`
})
class TestHostMatChipComponent {
}

@Component({
  selector: 'mat-paginator',
  template: `<div id="page" (click)="onPageChange($event)"></div>`
})
class TestHostMatPaginatorComponent {
  @Input('hidden') hidden: boolean;
  @Input('length') length: number;
  @Input('pageSize') pageSize: any;
  @Input('pageIndex') pageIndex: any;
  @Input('showFirstLastButtons') showFirstLastButtons: any;
  @Input('pageSizeOptions') pageSizeOptions: any;
  @Output() page = new EventEmitter();

  onPageChange(event) {
    this.page.emit(event.page);
  }
}


describe('HomeComponent', () => {
  let component: HomeComponent;
  let fixture: ComponentFixture<HomeComponent>;
  let de: DebugElement;
  let store: Store<fromRoot.AppState>;
  let profile: Profile;
  let user: User;
  let article: Article;

  const returnConfig: ArticlesConfigState = {
    type: 'all',
    filters: {
      tag: 'hello',
      author: '',
      favorited: '',
      offset: 0,
      limit: 10,
      pageIndex: 0
    }
  };

  beforeEach(
    async(() => {
      profile = getProfile();
      user = getUser();
      article = getArticle();
      const configure: ConfigureFn = testBed => {
        testBed.configureTestingModule({
          declarations: [
            HomeComponent,
            TestHostTagsComponent,
            TestHostArticleListComponent,
            TestHostMatPaginatorComponent,
            TestHostContentTabsComponent,
            TestHostMatChipComponent,
            ShowAuthedDirective
          ],
          imports: [
            NoopAnimationsModule,
            StoreModule.forRoot({
              ...fromRoot.reducers,
            }),
          ],
        });
      };

      configureTests(configure).then(testBed => {
        fixture = testBed.createComponent(HomeComponent);
        component = fixture.componentInstance;
        de = fixture.debugElement;
        store = testBed.get(Store);
        fixture.detectChanges();

        spyOn(store, 'dispatch').and.callThrough();
      });
    })
  );

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should compile', () => {
    (<any>expect(fixture)).toMatchSnapshot();
  });

  it('should show selected tag', () => {
    store.dispatch(new ArticlesActions.SetTag({ tag: 'tag' }));
    fixture.detectChanges();
    (<any>expect(fixture)).toMatchSnapshot();
  });

  it('should hide banner for logged in user', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    fixture.detectChanges();
    (<any>expect(fixture)).toMatchSnapshot();
  });


  it('should load articles for not loggen in user', () => {
    component.ngOnInit();

    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ClearReturnArticlesConfig());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadTagsRequest());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
  });

  it('should load articles for loggen in user', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    component.ngOnInit();

    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ClearReturnArticlesConfig());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetTypeOfFeed({ type: 'feed' }));
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadTagsRequest());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
  });

  it('should load articles for loggen in user and return config', () => {
    store.dispatch(new AuthActions.LoginSuccess({ user }));
    store.dispatch(new ArticlesActions.SetReturnArticlesConfig({ config: returnConfig }));
    component.ngOnInit();

    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ClearReturnArticlesConfig());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());

    const { tag, offset, limit } = returnConfig.filters;
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetTypeOfFeed({ type: returnConfig.type }));
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetLimit({ limit }));
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetOffset({ offset }));
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetPageIndex({ pageIndex: offset / limit }));
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetTag({ tag }));

    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadTagsRequest());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
  });

  it('should select tags', () => {
    component.ngOnInit();

    const expected = cold('a', { a: [] });

    expect(component.tags$).toBeObservable(expected);
  });

  it('should select currentTag', () => {
    component.ngOnInit();

    const expected = cold('a', { a: '' });

    expect(component.currentTag$).toBeObservable(expected);
  });

  it('should select loadingTags', () => {
    component.ngOnInit();

    const expected = cold('a', { a: true });

    expect(component.loadingTags$).toBeObservable(expected);
  });

  it('should select errorLoadingTags', () => {
    component.ngOnInit();

    const expected = cold('a', { a: false });

    expect(component.errorLoadingTags$).toBeObservable(expected);
  });

  it('should select type', (done) => {
    component.ngOnInit();

    component.type$.subscribe(res => {
      expect(res).toBe('all');
      done();
    }, done, done);
  });

  it('should select atricles', () => {
    component.ngOnInit();

    const expected = cold('a', { a: [] });

    expect(component.atricles$).toBeObservable(expected);
  });

  it('should select limit', () => {
    component.ngOnInit();

    const expected = cold('a', { a: 10 });

    expect(component.limit$).toBeObservable(expected);
  });

  it('should select pageIndex', () => {
    component.ngOnInit();

    const expected = cold('a', { a: 0 });

    expect(component.pageIndex$).toBeObservable(expected);
  });

  it('should select articlesCount', () => {
    component.ngOnInit();

    const expected = cold('a', { a: 0 });

    expect(component.articlesCount$).toBeObservable(expected);
  });

  it('should select loadingArticles', () => {
    component.ngOnInit();

    const expected = cold('a', { a: true });

    expect(component.loadingArticles$).toBeObservable(expected);
  });

  it('should dispatch "ToggleArticleFavoriteRequest"', () => {
    const clickDe = de.query(By.css('#favorite'));

    clickDe.triggerEventHandler('click', null);

    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ToggleArticleFavoriteRequest({ article }));
  });

  it('should dispatch "ToggleArticleFavoriteRequest"', () => {
    const clickDe = de.query(By.css('#selectTag'));

    clickDe.triggerEventHandler('click', null);

    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetTag({tag: 'tag'}));
    expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
  });

  it('should load article for sertain tab and set author', () => {
      const spy = jest.spyOn(store, 'pipe').mockImplementation(() => of(false));
      const tabDe = de.query(By.css('#tabs'));
      tabDe.triggerEventHandler('click', { type: 'feed' });

      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetTypeOfFeed({ type: 'feed' }));
      expect(store.dispatch).toHaveBeenCalledWith(new AuthActions.LogoutAction());
  });

  it('should load article for sertain tab and set favorited', () => {
    const spy = jest.spyOn(store, 'pipe').mockImplementation(() => of(true));
    const tabDe = de.query(By.css('#tabs'));
    tabDe.triggerEventHandler('click', { type: 'notfeed' });

      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.ResetConfig());
      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetTypeOfFeed({ type: 'notfeed' }));
      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
  });

  it('should change page', () => {
      const page = {
          pageIndex: 20,
          pageSize: 20
      };
      const pageDe = de.query(By.css('#page'));
      pageDe.triggerEventHandler('click', { page });

      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetLimit({ limit: page.pageSize }));
      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetPageIndex({ pageIndex: page.pageIndex }));
      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.SetOffset({ offset: (page.pageSize * page.pageIndex) }));
      expect(store.dispatch).toHaveBeenCalledWith(new ArticlesActions.LoadArticlesRequest());
  });
});
