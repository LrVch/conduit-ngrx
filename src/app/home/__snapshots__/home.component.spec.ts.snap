// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeComponent should compile 1`] = `
<ng-component
  articlesCount$={[Function Store]}
  atricles$={[Function Store]}
  currentTag$={[Function Store]}
  errorLoadingTags$={[Function Store]}
  limit$={[Function Store]}
  loadingArticles$={[Function Store]}
  loadingTags$={[Function Store]}
  locale$={[Function Store]}
  pageIndex$={[Function Store]}
  pageSizeOptions={[Function Array]}
  store={[Function Store]}
  tabs$={[Function Observable]}
  tags$={[Function Store]}
  translateService={[Function TranslateService]}
  type$={[Function Observable]}
>
  <section>
    
    <div
      class="banner text-center"
    >
      <div
        class="container"
      >
        <h1>
          conduit.home.baner.title
        </h1>
        <p>
          conduit.home.baner.caption.
        </p>
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-9 col-sm-12"
          >
            <app-content-tabs
              ng-reflect-tabs="[object Object],[object Object"
              ng-reflect-type="all"
            >
              <div
                id="tabs"
              >
                
              </div>
            </app-content-tabs>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row flex-column-reverse flex-lg-row"
        >
          <div
            class="col-lg-9 col-sm-12"
          >
            <div
              class="articles"
            >
              <app-articles-list
                ng-reflect-articles-list=""
                ng-reflect-content-length="10"
                ng-reflect-is-loading="true"
                ng-reflect-locale=""
              >
                <div
                  id="favorite"
                />
              </app-articles-list>
            </div>
          </div>
          <div
            class="col-lg-3 col-sm-12"
          >
            <div
              class="tags-wtapper"
            >
              <app-tags
                ng-reflect-current-tag=""
                ng-reflect-is-error-loading="false"
                ng-reflect-is-loading="true"
                ng-reflect-tags=""
              >
                <div
                  id="selectTag"
                />
              </app-tags>
            </div>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-9 col-sm-12"
          >
            <mat-paginator
              ng-reflect-hidden="true"
              ng-reflect-length="0"
              ng-reflect-page-index="0"
              ng-reflect-page-size="10"
              ng-reflect-page-size-options="5,10,25,100"
              ng-reflect-show-first-last-buttons="true"
            >
              <div
                id="page"
              />
            </mat-paginator>
          </div>
        </div>
        <div
          class="pagination"
        />
      </div>
    </div>
  </section>
</ng-component>
`;

exports[`HomeComponent should hide banner for logged in user 1`] = `
<ng-component
  articlesCount$={[Function Store]}
  atricles$={[Function Store]}
  currentTag$={[Function Store]}
  errorLoadingTags$={[Function Store]}
  limit$={[Function Store]}
  loadingArticles$={[Function Store]}
  loadingTags$={[Function Store]}
  locale$={[Function Store]}
  pageIndex$={[Function Store]}
  pageSizeOptions={[Function Array]}
  store={[Function Store]}
  tabs$={[Function Observable]}
  tags$={[Function Store]}
  translateService={[Function TranslateService]}
  type$={[Function Observable]}
>
  <section>
    
    <div
      class="content"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-9 col-sm-12"
          >
            <app-content-tabs
              ng-reflect-tabs="[object Object],[object Object"
              ng-reflect-type="all"
            >
              <div
                id="tabs"
              >
                
              </div>
            </app-content-tabs>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row flex-column-reverse flex-lg-row"
        >
          <div
            class="col-lg-9 col-sm-12"
          >
            <div
              class="articles"
            >
              <app-articles-list
                ng-reflect-articles-list=""
                ng-reflect-content-length="10"
                ng-reflect-is-loading="true"
                ng-reflect-locale=""
              >
                <div
                  id="favorite"
                />
              </app-articles-list>
            </div>
          </div>
          <div
            class="col-lg-3 col-sm-12"
          >
            <div
              class="tags-wtapper"
            >
              <app-tags
                ng-reflect-current-tag=""
                ng-reflect-is-error-loading="false"
                ng-reflect-is-loading="true"
                ng-reflect-tags=""
              >
                <div
                  id="selectTag"
                />
              </app-tags>
            </div>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-9 col-sm-12"
          >
            <mat-paginator
              ng-reflect-hidden="true"
              ng-reflect-length="0"
              ng-reflect-page-index="0"
              ng-reflect-page-size="10"
              ng-reflect-page-size-options="5,10,25,100"
              ng-reflect-show-first-last-buttons="true"
            >
              <div
                id="page"
              />
            </mat-paginator>
          </div>
        </div>
        <div
          class="pagination"
        />
      </div>
    </div>
  </section>
</ng-component>
`;

exports[`HomeComponent should show selected tag 1`] = `
<ng-component
  articlesCount$={[Function Store]}
  atricles$={[Function Store]}
  currentTag$={[Function Store]}
  errorLoadingTags$={[Function Store]}
  limit$={[Function Store]}
  loadingArticles$={[Function Store]}
  loadingTags$={[Function Store]}
  locale$={[Function Store]}
  pageIndex$={[Function Store]}
  pageSizeOptions={[Function Array]}
  store={[Function Store]}
  tabs$={[Function Observable]}
  tags$={[Function Store]}
  translateService={[Function TranslateService]}
  type$={[Function Observable]}
>
  <section>
    
    <div
      class="banner text-center"
    >
      <div
        class="container"
      >
        <h1>
          conduit.home.baner.title
        </h1>
        <p>
          conduit.home.baner.caption.
        </p>
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-9 col-sm-12"
          >
            <app-content-tabs
              ng-reflect-tabs="[object Object],[object Object"
              ng-reflect-type=""
            >
              <div
                id="tabs"
              >
                
                <mat-chip
                  color="primary"
                  selected=""
                >
                  #tag
                </mat-chip>
              </div>
            </app-content-tabs>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row flex-column-reverse flex-lg-row"
        >
          <div
            class="col-lg-9 col-sm-12"
          >
            <div
              class="articles"
            >
              <app-articles-list
                ng-reflect-articles-list=""
                ng-reflect-content-length="10"
                ng-reflect-is-loading="true"
                ng-reflect-locale=""
              >
                <div
                  id="favorite"
                />
              </app-articles-list>
            </div>
          </div>
          <div
            class="col-lg-3 col-sm-12"
          >
            <div
              class="tags-wtapper"
            >
              <app-tags
                ng-reflect-current-tag="tag"
                ng-reflect-is-error-loading="false"
                ng-reflect-is-loading="true"
                ng-reflect-tags=""
              >
                <div
                  id="selectTag"
                />
              </app-tags>
            </div>
          </div>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="col-md-9 col-sm-12"
          >
            <mat-paginator
              ng-reflect-hidden="true"
              ng-reflect-length="0"
              ng-reflect-page-index="0"
              ng-reflect-page-size="10"
              ng-reflect-page-size-options="5,10,25,100"
              ng-reflect-show-first-last-buttons="true"
            >
              <div
                id="page"
              />
            </mat-paginator>
          </div>
        </div>
        <div
          class="pagination"
        />
      </div>
    </div>
  </section>
</ng-component>
`;
